prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala

```diff
@@ -179,15 +183,68 @@ trait Functions extends Base with Operators with Vectors {
             timeParams.start * 1000, timeParams.step * 1000, timeParams.end * 1000, 0,
             rangeFunctionId, false, otherParams, instantExpression.offset.map(_.millis(timeParams.step * 1000)))
         } else {
-          val rangeExpression = seriesParam.asInstanceOf[RangeExpression]
-          PeriodicSeriesWithWindowing(
-            rangeExpression.toSeriesPlan(timeParams, isRoot = false),
-            timeParams.start * 1000 , timeParams.step * 1000, timeParams.end * 1000,
-            rangeExpression.window.millis(timeParams.step * 1000),
-            rangeFunctionId, rangeExpression.window.timeUnit == IntervalMultiple,
-            otherParams, rangeExpression.offset.map(_.millis(timeParams.step * 1000)))
+          seriesParam match {
+            case re: RangeExpression => rangeExpressionArgument(re, timeParams, rangeFunctionId, otherParams)
+            case sq: SubqueryExpression => subqueryArgument(sq, timeParams, rangeFunctionId, otherParams)
+          }
         }
       }
     }
+
+    def rangeExpressionArgument(
+      rangeExpression : RangeExpression,
+      timeParams: TimeRangeParams,
+      rangeFunctionId: RangeFunctionId,
+      otherParams: Seq[FunctionArgsPlan]
+    ) : PeriodicSeriesPlan = {
+      PeriodicSeriesWithWindowing(
+        rangeExpression.toSeriesPlan(timeParams, isRoot = false),
+        timeParams.start * 1000, timeParams.step * 1000, timeParams.end * 1000,
+        rangeExpression.window.millis(timeParams.step * 1000),
+        rangeFunctionId,
+        rangeExpression.window.timeUnit == IntervalMultiple,
+        otherParams,
+        rangeExpression.offset.map(_.millis(timeParams.step * 1000))
+      )
+    }
+
+    def subqueryArgument(
+      sqe : SubqueryExpression,
+      timeParams: TimeRangeParams,
+      rangeFunctionId: RangeFunctionId,
+      otherParams: Seq[FunctionArgsPlan]
+    ) : PeriodicSeriesPlan = {
+      var subqueryStepToUseMs = 60000L
+      if (sqe.sqcl.step.isDefined) {
+        subqueryStepToUseMs = sqe.sqcl.step.get.millis(1L)
+      }
+      // when start and stop are the same, step should be zero too
+      var outerStepMs = timeParams.step * 1000
+      if (timeParams.start == timeParams.end) {
+        outerStepMs = 0
+      }
+
+      // We don't want to naively execute subquery by concatenating
+      // the results of multiple queries. Instead, the inner subqueries are executed
+      // as a wider single query with newly calculated start and step. Later we
+      // apply a transformation against the resulting time series to form the actual
+      // response of the subquery.
+      @tailrec def gcd(a: Long, b: Long): Long = if (b == 0) a else gcd(b, a % b)
+      val newStepForInner = gcd(outerStepMs, subqueryStepToUseMs)
+
+      val timeParamsForInner = TimeStepParams(
+        timeParams.start - (sqe.sqcl.window.millis(1L) / 1000),
+        newStepForInner/1000,
+        timeParams.end
+      )
```

Comment:
Why are execution decisions being made at the logical plan construction ? This should be done at the ExecPlan materialization stage. LogicalPlan should be simple representation of the query as pojos. Thats it.